flink-tests/src/test/java/org/apache/flink/test/scheduling/PipelinedRegionSchedulingITCase.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.flink.client.program.MiniClusterClient;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.JobManagerOptions;
+import org.apache.flink.configuration.RestOptions;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionType;
+import org.apache.flink.runtime.jobgraph.DistributionPattern;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.jobgraph.JobVertex;
+import org.apache.flink.runtime.jobgraph.ScheduleMode;
+import org.apache.flink.runtime.jobmanager.scheduler.NoResourceAvailableException;
+import org.apache.flink.runtime.jobmanager.scheduler.SlotSharingGroup;
+import org.apache.flink.runtime.jobmaster.JobResult;
+import org.apache.flink.runtime.minicluster.MiniCluster;
+import org.apache.flink.runtime.minicluster.MiniClusterConfiguration;
+import org.apache.flink.runtime.testtasks.NoOpInvokable;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+/**
+ * IT case for pipelined region scheduling.
+ */
+public class PipelinedRegionSchedulingITCase extends TestLogger {
+
+	@Test
+	public void testSuccessWithSlotsNoFewerThanTheMaxRegionRequired() throws Exception {
+		final JobResult jobResult = executeSchedulingTest(2);
+		assertThat(jobResult.getSerializedThrowable().isPresent(), is(false));
+	}
+
+	@Test
+	public void testFailsOnInsufficientSlots() throws Exception {
+		final JobResult jobResult = executeSchedulingTest(1);
+		assertThat(jobResult.getSerializedThrowable().isPresent(), is(true));
+
+		final Throwable jobFailure = jobResult
+			.getSerializedThrowable()
+			.get()
+			.deserializeError(ClassLoader.getSystemClassLoader());
+
+		final Optional<NoResourceAvailableException> cause = ExceptionUtils.findThrowable(
+			jobFailure,
+			NoResourceAvailableException.class);
+		assertThat(cause.isPresent(), is(true));
+		assertThat(cause.get().getMessage(), containsString("Slot request bulk is not fulfillable!"));
+	}
+
+	private JobResult executeSchedulingTest(int numSlots) throws Exception {
+		final Configuration configuration = new Configuration();
+		configuration.setString(RestOptions.BIND_PORT, "0");
+		configuration.setLong(JobManagerOptions.SLOT_REQUEST_TIMEOUT, 5000L);
+
+		final MiniClusterConfiguration miniClusterConfiguration = new MiniClusterConfiguration.Builder()
+			.setConfiguration(configuration)
+			.setNumTaskManagers(1)
+			.setNumSlotsPerTaskManager(numSlots)
+			.build();
+
+		try (MiniCluster miniCluster = new MiniCluster(miniClusterConfiguration)) {
+			miniCluster.start();
+
+			final MiniClusterClient miniClusterClient = new MiniClusterClient(configuration, miniCluster);
+
+			final JobGraph jobGraph = createJobGraph(100);
```

Comment:
No big difference. Just that 100 better proves that one physical slot can be reused for different shared slot for multiple times.
The parallelism does not significantly increase the execution time so I think it's not bad to have it.

Reply:
I have improved the testing job to not pass with lazy-form-sources scheduling. The parallelism is decreased to 10.
A parallelism > 8 is required here. This is because in lazy-form-sources scheduling, the downstream task will wait for all the upstream task to be assigned slots if the parallelism is very small. This will result in source tasks to be scheduled earlier than sink tasks, and the job will not deadlock due to this side effect.